File: 26. Find All Anagrams in a String.cpp
// Topic: Find All Anagrams in a String
#include <bits/stdc++.h>
using namespace std;

vector<int> findAnagrams(string s, string p)
{
    vector<int> res;
    if (p.size() > s.size())
        return res;
    vector<int> cnt1(26, 0), cnt2(26, 0);
    for (int i = 0; i < p.size(); ++i)
    {
        cnt2[p[i] - 'a']++;
        cnt1[s[i] - 'a']++;
    }
    if (cnt1 == cnt2)
        res.push_back(0);
    for (int i = p.size(); i < s.size(); ++i)
    {
        cnt1[s[i] - 'a']++;
        cnt1[s[i - p.size()] - 'a']--;
        if (cnt1 == cnt2)
            res.push_back(i - p.size() + 1);
    }
    return res;
}

int main()
{
    string s = "cbaebabacd", p = "abc";
    auto idx = findAnagrams(s, p);
    cout << "Indices:";
    for (int i : idx) cout << " " << i;
    cout << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 26. Find All Anagrams in a String.py
# Topic: Find All Anagrams in a String

def find_anagrams(s, p):
    res = []
    if len(p) > len(s): return res
    cnt1 = [0]*26
    cnt2 = [0]*26
    for i in range(len(p)):
        cnt1[ord(s[i]) - ord('a')] += 1
        cnt2[ord(p[i]) - ord('a')] += 1
    if cnt1 == cnt2:
        res.append(0)
    for i in range(len(p), len(s)):
        cnt1[ord(s[i]) - ord('a')] += 1
        cnt1[ord(s[i-len(p)]) - ord('a')] -= 1
        if cnt1 == cnt2:
            res.append(i-len(p)+1)
    return res

if __name__ == '__main__':
    s = 'cbaebabacd'
    p = 'abc'
    print('Indices:', find_anagrams(s,p))

# Time Complexity: O(n)
# Space Complexity: O(1)
